staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -1011,6 +1011,10 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 
 	handler = genericfilters.WithCORS(handler, c.CorsAllowedOriginList, nil, nil, nil, "true")
 
+	// WithWarningRecorder must be wrapped by the timeout handler
+	// to make the addition of warning headers threadsafe
+	handler = genericapifilters.WithWarningRecorder(handler)
```

Comment:
actually, thinking through this more, just moving the warning recorder doesn't fix the panic

the warning recorder does things like this:

```
r.writer.Header().Add("Warning", header)
```

`baseTimeoutWriter#Header()` is threadsafe, but once it returns the http.Header object, the call to Add() is *not* threadsafe. That means in a timeout case, baseTimeoutWriter#Write → copyHeaders() can still be trying to read from tw.handlerHeaders at the same time the warning recorder is calling Add() and mutating the headers.

I don't think this PR actually makes the warning header additions threadsafe. #123034 wraps the Add/Set/Del calls to make *those* threadsafe.

Reply:
after the request times out, the timeout handler returns an empty `http.Header`

https://github.com/kubernetes/kubernetes/blob/a9e4f5b7862e84c4152eabe2e960f3f6fb9a4867/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L191-L193

and this panic seems to manifests only after a request times out. Maybe we can add a test to demonstrate that the panic does not exhibit before the request times out, thoughts?

Reply:
oh, in the timeout case, baseTimeoutWriter#Write never makes it to the copyHeaders call and we just discard anything the warning recorded added to tw.handlerHeaders

this works

Reply:
thanks, nevertheless, i will try to work on a unit test that reproduces the `concurrent map access` panic

Reply:
I think the diff at https://github.com/kubernetes/kubernetes/pull/115282#discussion_r1473107715 will catch that